File: python/sliding_window/longest_repeating_char_replacement.py

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        counts = [0] * 26
        left = 0
        max_freq = 0
        best = 0

        for right, char in enumerate(s):
            idx = ord(char) - ord("A")
            counts[idx] += 1
            max_freq = max(max_freq, counts[idx])

            while right - left + 1 - max_freq > k:
                counts[ord(s[left]) - ord("A")] -= 1
                left += 1

            best = max(best, right - left + 1)

        return best

File: typescript/src/sliding-window/longest-repeating-char-replacement.ts
export class LongestRepeatingCharReplacement {
  characterReplacement(s: string, k: number): number {
    const counts = new Array<number>(26).fill(0);
    let left = 0;
    let maxFreq = 0;
    let best = 0;

    for (let right = 0; right < s.length; right++) {
      const idx = s.charCodeAt(right) - "A".charCodeAt(0);
      counts[idx]++;
      maxFreq = Math.max(maxFreq, counts[idx]);

      while (right - left + 1 - maxFreq > k) {
        counts[s.charCodeAt(left) - "A".charCodeAt(0)]--;
        left++;
      }

      best = Math.max(best, right - left + 1);
    }

    return best;
  }
}
